### PR TITLE
Avoid retrying responses with 404 status code

### PR DESCRIPTION
**Pull request by mouradmourafiq**
related #405

### CODE REVIEW

File: gcsfs/retry.py [hunk 1]
```diff
@@ -121,8 +111,15 @@ async def retry_request(func, retries=6, *args, **kwargs):
...
-                msg = "Bucket is requester pays. Set `requester_pays=True` when creating the GCSFileSystem."
+                msg = (
+                    "Bucket is requester pays. "
+                    "Set `requester_pays=True` when creating the GCSFileSystem."
+                )
                 raise ValueError(msg) from e
+            # Special test for 404 to avoid retrying the request
+            if isinstance(e, aiohttp.client_exceptions.ClientError) and e.code == 404:
+                logger.debug("Request returned 404, no retries.")
+                raise e
```
Comment (martindurant) [inline gcsfs/retry.py:132]: Just `raise` alone

**Review by martindurant — COMMENTED** [review]
I don't have a principled object to changing around the codes and messaging here. I wish we could get someone with a deeper understanding of GCS's errors and http/aiohttp in general to give their opinions.

File: gcsfs/retry.py (hunk 1 above)
Comment (md-mq) [reply gcsfs/retry.py:132]: Just to be consistent with the rest of the file, see this [line](https://github.com/dask/gcsfs/pull/406/files#diff-d89213d3b1e29d457dc7aed940ccb379f48e8ef45f34495e2ab3934e39124534L130) and this [line](https://github.com/dask/gcsfs/pull/406/files#diff-d89213d3b1e29d457dc7aed940ccb379f48e8ef45f34495e2ab3934e39124534L135)

**Comment by mouradmourafiq** [comment]
Given that the tests are failing, should I revert the change related to the [json loading of the error](https://github.com/dask/gcsfs/pull/406/files#diff-d89213d3b1e29d457dc7aed940ccb379f48e8ef45f34495e2ab3934e39124534L83) or should I adapt the tests to the new behavior.

Just to give more information about why I made the change, in my tests with large folders with several nesting levels and files, when errors are raised, `json.loads` is a blocking behavior and impacts performance, additionally the error formatting does not bring any value in terms of debugging, keeping the full content users can decide to handle the errors later for debugging purposes, but most users are will trade that for the performance benefits.

**Comment by martindurant** [comment]
> should I revert the change related to the json loading of the error

If this leads to tests passing locally, then probably yes. Again, the JSON thing could be a new PR, but I'm not convinced it's a problem - decoding should always take a trivial amount of time.

File: gcsfs/retry.py [hunk 2]
```diff
@@ -76,6 +76,9 @@ def validate_response(status, content, path, headers=None):
     path: associated URL path, for error messages
     """
     if status >= 400:
+        if status == 404:
+            raise FileNotFoundError
```
Comment (martindurant) [inline gcsfs/retry.py]: Could we have `FileNotFoundError(path)` ?

Comment (md-mq) [reply gcsfs/retry.py]: done

**Comment by martindurant** [comment]
Thanks @mouradmourafiq

**Pull request merged**